02_knight_and_princess/robert_gaisin/main.cpp

```diff
@@ -0,0 +1,235 @@
+#include <ctime>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+using VertexId = int;
+using EdgeId = int;
+
+std::ostream& operator<<(std::ostream& out, const vector<int>& int_vector) {
+  for (auto it = int_vector.begin(); it != int_vector.end(); ++it) {
+    out << *it;
+    if (it != int_vector.end() - 1) {
+      out << ", ";
+    }
+  }
+  return out;
+}
+
+struct Vertex {
+  VertexId id = 0;
+  int depth = 0;
+  vector<EdgeId> edge_ids;
+};
+
+std::ostream& operator<<(std::ostream& out, const Vertex& vertex) {
+  out << "{" << endl;
+  out << "      \"id\": " << vertex.id << "," << endl;
+  out << "      \"edge_ids\": [" << vertex.edge_ids << "]" << endl;
+  out << "    }";
+  return out;
+}
+
+struct Edge {
+  EdgeId id = 0;
+  VertexId begin = 0;
+  VertexId end = 0;
+};
+
+std::ostream& operator<<(std::ostream& out, const Edge& edge) {
+  out << "{" << endl;
+  out << "      \"id\": " << edge.id << "," << endl;
+  out << "      \"vertex_ids\": [" << edge.begin << ", " << edge.end << "]"
+      << endl;
+  out << "    }";
+  return out;
+}
+
+class Graph {
+ public:
+  void resize(int graph_depth) { vertices_.resize(graph_depth); }
+  VertexId next_vertex_id() { return num_of_vrt_++; }
+  EdgeId next_edge_id() { return num_of_edg_++; }
+  void add_vertex(Vertex& vertex, int layer);
+  void add_edge(Edge& edge) { edges_.push_back(edge); }
+  void add_edge_in_vertex(int layer, int vertex_in_layer, EdgeId edge_id);
+  bool layer_is_empty_or_not_exist(int layer);
+  VertexId get_vertex_id(int layer, int vertex_in_layer);
+  size_t layer_size(int layer);
+
+  const vector<vector<Vertex>> vertices() const { return vertices_; }
+  const vector<Edge> edges() const { return edges_; }
+
+ private:
+  int depth_ = 0;
+  VertexId num_of_vrt_ = 0;
+  EdgeId num_of_edg_ = 0;
+  vector<vector<Vertex>> vertices_;
```

Comment:
Для реализации логики графа из первой задачи, необходимости усложнять и добавлять "глубину" особо нет.
Хорошей практикой считается, закладывать удобно масштабируемую архитектуру, но при этом писать как можно меньше кода, и добавлять как можно меньше логики. Любое усложнение - потенциальная ошибка в будущем. Да и ревью мне делать значительно сложнее, что так же увеличивает время принятия ПР.
```suggestion
  vector<Vertex> vertices_;
```